src/modules/kvs/commit.c

```diff
@@ -268,21 +278,25 @@ static int commit_unroll (commit_t *c, int current_epoch, json_t *dir)
                     return -1;
                 }
             }
-            if (!(tmpdirent = j_dirent_create ("DIRREF", ref)))
+            if (!(tmp = treeobj_create_dirref (ref)))
                 return -1;
-            if (json_object_iter_set_new (dir, iter, tmpdirent) < 0) {
-                json_decref (tmpdirent);
+            if (json_object_iter_set_new (dir, iter, tmp) < 0) {
+                json_decref (tmp);
                 errno = ENOMEM;
                 return -1;
             }
         }
-        else if ((key_value = json_object_get (value, "FILEVAL"))) {
+        else if (treeobj_is_val (dir_entry)) {
+            json_t *val_data;
             size_t size;
-            if (kvs_util_json_encoded_size (key_value, &size) < 0)
+
+            if (!(val_data = treeobj_get_data (dir_entry)))
+                return -1;
+            if (kvs_util_json_encoded_size (val_data, &size) < 0)
                 return -1;
             if (size > BLOBREF_MAX_STRING_SIZE) {
-                json_incref (key_value);
-                if ((ret = store_cache (c, current_epoch, key_value,
+                json_incref (val_data);
+                if ((ret = store_cache (c, current_epoch, val_data,
```

Comment:
Two out of three uses of `store_cache()` call `json_incref()` on the third argument before passing it in.  Maybe it would be a bit simpler if `store_cache()` didn't "steal" the reference?